### ISSUE TITLE
Proposal: changes to client message protocols

### ISSUE THREAD

**Issue by yilunzhang** [nknorg/nkn-protocols#1]
The current client message protocol contains some parts that was intended to be backward compatible but make the protocol a little bit complex. We are having some discussion recently to deprecate some fields and make it simpler and less confusing.

For the client-to-client message protocol (https://github.com/nknorg/nkn-client-js/blob/master/lib/protocol/payloads.proto):
* Deprecate TEXT in PayloadType;
* Deprecate TextData message type;
* Basically message is only in binary format. To send text message, sender serializes the text into binary format in some way, and receiver deserializes the binary message back into text.

For the client-to-node message protocol (https://github.com/nknorg/nkn/blob/master/pb/clientmessage.proto):
* Deprecate dest in OutboundMessage and uses dests only
* Rename OutboundMessage to ClientToNodeMessage
* Rename InboundMessage to NodeToClientMessage

Please let me know how you think about it. @trueinsider @RealJohnSmith @termoose

**Comment by trueinsider** [comment]
I think we can deprecate `type` too, because `ACK` is basically `BINARY` with empty `data`.

Also, I think we can discuss here changes to support encryption and sigchain too.

**Comment by yilunzhang** [comment]
I think it's still good to keep this PayloadType layer in case we need more types in the future.

Also I think it's better to separate ACK from empty response, otherwise we kind of force response to be non-empty from the protocol layer, which I think is not good for separating protocol and content in the application layer.

**Comment by trueinsider** [comment]
I think it can be kinda confusing having to differentiate between `ACK empty response` and `BINARY empty response` at dApp level. And I can't think of a use-case where this would be actually needed.

**Comment by yilunzhang** [comment]
I don't think dApp developer needs to deal with (or even aware of) ACK response -- it should be done automatically by SDK and hide from the developer like the JS SDK is doing now.

In addition, empty and empty are different. ACK vs empty response is like null vs empty array in JS, they are different type and have different meaning.

Having an ACK type has a few advantages IMO:
1. We can have different policies for ACK and response, e.g. whether to allow empty msg id, node caching msg, etc.
2. We can include non-empty content in ACK packet from the protocol level in the future (ACK does not have to be empty)
3. We can allow empty response to be sent (BINARY can be empty)

**Comment by trueinsider** [comment]
1. node don't know anything about this protocol level and probably shouldn't
2-3. yeah, and what I'm saying that I don't see any real use case for it. you can do this, sure, but why would you do it?

**Comment by yilunzhang** [comment]
1. I agree, and I shouldn't use node caching msg as example. But even for the client level it could still be true.

For 2 and 3, TBH I'm not sure either, but I think it's always good to give developer the flexibility as some of them may want to use empty response for special purpose, especially given that we don't sacrifice efficiency or introduce extra complexity for them (since ACK is hidden from developer).

**Comment by termoose** [comment]
I agree we should not use empty `BINARY` as `ACK`, somewhere in the future there will be a developer using empty messages for something and it will be handled differently than a standard `BINARY` causing a debugging nightmare :)

I'm curious about having messages only in binary format, it seems like it should work well and I found a discussion on it [here](https://groups.google.com/forum/#!topic/protobuf/au6eQBkRT5s).
Judging from the discussion there are some protobuf libraries that give different types if the data comes from `bytes` instead of `string`. In Java you get `ByteString` for `bytes` for example. I'm not sure if this is a problem? The SDK in the respective languages could have conversion functions/flags for this where the user explicitly asks for a normal `String` and it's converted automagically. What happens if it can't convert (crash, exception)? Sounds like we lose some type safety from protobuf by doing this.

**Comment by RealJohnSmith** [comment]
> The current client message protocol contains some parts that was intended to be backward compatible

I wouldn't worry about backward compatibility at all since we are only on testnet (and SDKs can hide protocol changes from app developers anyway)


> The SDK in the respective languages could have conversion functions/flags for this where the user explicitly asks for a normal `String` and it's converted automagically.


The problem about language dependent byte/string conversions is that if you send a message between, let's say, Java and Javascript client, it may be encoded/decoded differently (especially some edge cases). I think SDKs should provide a standard way of converting bytes to string and back (That doesn't mean, that there has to be a packet type for it. More like utility method if you want to send both binary and string data in one message)


I would appreciate more some support for signing and encryption at a protocol level.
Let's say, add a message type `KEY_EXCHANGE` which would be completely hidden from user/dApp developer and would be used only by the client when starting communication with another client, to establish a symetric key for further communication (and encrypt/decrypt 
 and sign/verify everything in the background)

**Comment by yilunzhang** [comment]
I also think language independent serialization/deseralization would be good, that’s why we have text message type currently. I’m not too worried about cross language compatibility if we switch to bytes only because there are quite a few mature language neutral ways to serialize or deserialize strings into and from bytes. So personally both solutions sound acceptable to me and we should probably discuss more about it here to reach a consensus.

For the encryption I definitely agree we should have some protocol level specs and implementions in SDK (and probably should be used by default) so developers just need to choose whether they want to use it. But I was thinking we should use a separate issue to discuss about it.

**Comment by RealJohnSmith** [comment]
I think encryption/signatures should be discussed in this issue as well because it is a part of the protocol, and we might need to add some fields/packet types and stuff like that.

When it comes to binary/text data type, I would do binary only, with SDK level utilities to encode/decode data (or only a part) between different common formats (string, object as JSON, ...)
This way it would be possible to combine multiple encodings (or just use one) in a single message, if the application calls for it. 
`ACK` should be something completely different than `BINARY` message without content, because user/developer is not supposed to interact with `ACK` messages but is supposed to do whatever he wants with `BINARY` messages (or rather, `a message` from their perspective. Not even BIN/TEXT)

**Comment by yilunzhang** [comment]
This whole repo is for nkn protocol (node, client, etc) so we should have many issues in the future for client protocols :) Just in case you didn't see it, I created an issue a few hours ago for client encryption and signature: https://github.com/nknorg/nkn-protocols/issues/2
